### PR TITLE
feat: add cache middleware

### LINKED ISSUE #319: ✨ support cache extension
## Is your feature request related to a problem? Please describe.

- Elegantly support cache

## Additional context

refer to:

- https://github.com/gin-contrib/cache
- https://github.com/gofiber/fiber/tree/master/middleware/cache

### PR DESCRIPTION
**Pull request by BaiZe1998**
#### What type of PR is this?

feat

#### What this PR does / why we need it (en: English/zh: Chinese):

en: add cache middleware for hertz
zh: 为 hertz 增加 cache 中间件

#### Which issue(s) this PR fixes:

Fixes https://github.com/cloudwego/hertz/issues/319

### CODE REVIEW

File: cache.go [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+	CacheStore persist.CacheStore
+
+	// CacheDuration
+	CacheDuration time.Duration
+}
+
+// GetCacheStrategyByRequest User can use this function to design custom cache strategy by request.
+// The first return value bool means whether this request should be cached.
+// The second return value Strategy determine the special strategy by this request.
+type GetCacheStrategyByRequest func(c *app.RequestContext) (bool, Strategy)
```
Comment (li-jin-gou) [inline cache.go]: use func(ctx context.Context , c *app.RequestContext)

File: cache.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+	return cache(defaultCacheStore, defaultExpire, options)
+}
+
+func cache(
+	defaultCacheStore persist.CacheStore,
+	defaultExpire time.Duration,
+	options *Options,
+) app.HandlerFunc {
+	if options.getCacheStrategyByRequest == nil {
+		panic("cache strategy is nil")
```
Comment (li-jin-gou) [inline cache.go]: use const or common err

File: cache.go [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+		{
+			respCache := &ResponseCache{}
+			err := cacheStore.Get(cacheKey, &respCache)
+			if err == nil {
+				replyWithCache(c, options, respCache)
+				options.hitCacheCallback(c)
+				return
+			}
+
+			if err != persist.ErrCacheMiss {
```
Comment (li-jin-gou) [inline cache.go]: use errors.Is

File: cache.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+				options.logger.Errorf("get cache error: %s, cache key: %s", err, cacheKey)
```
Comment (li-jin-gou) [inline cache.go]: use hlog for all logger

File: option.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+import (
+	"time"
+
+	"github.com/cloudwego/hertz/pkg/app"
+)
+
+// Options contains all options
+type Options struct {
+	logger Logger
```
Comment (li-jin-gou) [inline option.go]: delete it and use hlog

File: option.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+	}
+}
+
+// Logger define the logger interface
+type Logger interface {
+	Errorf(string, ...interface{})
+}
+
+// Discard the default logger that will discard all logs of cache
+type Discard struct{}
```
Comment (li-jin-gou) [inline option.go]: use hlog

File: option.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+func WithCacheStrategyByRequest(getGetCacheStrategyByRequest GetCacheStrategyByRequest) Option {
+	return Option{
+		F: func(o *Options) {
+			o.getCacheStrategyByRequest = getGetCacheStrategyByRequest
+		},
+	}
+}
+
+// OnHitCacheCallback define the callback when use cache
+type OnHitCacheCallback func(c *app.RequestContext)
```
Comment (li-jin-gou) [inline option.go]: add context.Context

File: option.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+var defaultHitCacheCallback = func(c *app.RequestContext) {}
```
Comment (li-jin-gou) [inline option.go]: add context.Context

File: option.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+func WithOnHitCache(cb OnHitCacheCallback) Option {
+	return Option{
+		F: func(o *Options) {
+			o.hitCacheCallback = cb
+		},
+	}
+}
+
+// OnMissCacheCallback define the callback when use cache
+type OnMissCacheCallback func(c *app.RequestContext)
```
Comment (li-jin-gou) [inline option.go]: add context.Context

File: persist/cache.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	"time"
+)
+
+// ErrCacheMiss represent the cache key does not exist in the store
+var ErrCacheMiss = errors.New("persist cache miss error")
+
+// CacheStore is the interface of a Cache backend
+type CacheStore interface {
+	// Get retrieves an item from the Cache. if key does not exist in the store, return ErrCacheMiss
+	Get(key string, value interface{}) error
```
Comment (li-jin-gou) [inline persist/cache.go]: Should we pass context.Context?

File: persist/redis.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+)
+
+// RedisStore store http response in redis
+type RedisStore struct {
```
Comment (li-jin-gou) [inline persist/redis.go:52]: add redis test and you need to use docker

File: licenses/LICENSE-REDIS [hunk 12]
```diff
@@ -0,0 +1,25 @@
+Copyright (c) 2013 The github.com/go-redis/redis Authors.
```
Comment (li-jin-gou) [inline licenses/LICENSE-REDIS:1]: add license https://github.com/chenyahui/gin-cache

**Review by li-jin-gou — CHANGES_REQUESTED** [review]

File: cache.go (hunk 1 above)
Comment (BaiZe1998) [reply cache.go]: i got it, thx

File: cache.go (hunk 2 above)
Comment (BaiZe1998) [reply cache.go]: i got it, thx

File: cache.go (hunk 3 above)
Comment (BaiZe1998) [reply cache.go]: i got it, thx

File: cache.go (hunk 4 above)
Comment (BaiZe1998) [reply cache.go]: i got it, thx

File: option.go (hunk 6 above)
Comment (BaiZe1998) [reply option.go]: i got it, thx

File: option.go (hunk 7 above)
Comment (BaiZe1998) [reply option.go]: i got it, thx

File: option.go (hunk 8 above)
Comment (BaiZe1998) [reply option.go]: i got it, thx

File: option.go (hunk 9 above)
Comment (BaiZe1998) [reply option.go]: i got it, thx

File: persist/redis.go (hunk 11 above)
Comment (BaiZe1998) [reply persist/redis.go:52]: i got it, thx

File: licenses/LICENSE-REDIS (hunk 12 above)
Comment (BaiZe1998) [reply licenses/LICENSE-REDIS:1]: i got it, thx

File: persist/cache.go (hunk 10 above)
Comment (BaiZe1998) [reply persist/cache.go]: i got it, thx

**Review by li-jin-gou — APPROVED** [review]

**Pull request merged**